models/migrations/v112.go

```diff
@@ -0,0 +1,41 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"os"
+
+	"code.gitea.io/gitea/models"
+	"xorm.io/builder"
+	"xorm.io/xorm"
+)
+
+func removeAttachmentMissedRepo(x *xorm.Engine) error {
+	type Attachment struct {
+		UUID string `xorm:"uuid"`
+	}
+	var start int
+	attachments := make([]*Attachment, 0, 50)
+	for {
+		err := x.Select("uuid").Where(builder.NotIn("release_id", builder.Select("id").From("`release`"))).
+			OrderBy("id").Limit(50, start).Find(&attachments)
+		if err != nil {
+			return err
+		}
+
+		for i := 0; i < len(attachments); i++ {
+			os.RemoveAll(models.AttachmentLocalPath(attachments[i].UUID))
+		}
+
+		if len(attachments) < 50 {
+			break
+		}
+		start += 50
+		attachments = attachments[:0]
+	}
+
+	_, err := x.Exec("DELETE FROM attachment WHERE release_id NOT IN (SELECT id FROM `release`)")
```

Comment:
I think it's better to delete and commit inside the loop, after the  `for{}` with `os.RemoveAll`, just in case the operation is lengthy. In that case `OrderBy` and the `start` parameter will not be needed.

Reply:
@guillep2k That will spend more time than current implementation. And what's the advantage of that implementation?

Reply:
@lunny It would just proceed in small batches, so if any problems come up during the procedure, at least the work that was done will be commited. On second thought, it not that important in this case.